### DISCUSSION TITLE
Data polling options in v4.0.0-rc.10

### DISCUSSION THREAD

**Discussion by kemeris2000** [lowdefy/lowdefy#1662] [Questions and Help]
What are options for data polling in v4.0.0-rc.10 without creating custom plugin? There is good [discussion](https://github.com/lowdefy/lowdefy/discussions/878) for v3.22, but as I understand JsAction has been removed in v4?

**Comment by Gervwyk** [comment]
Hi @kemeris2000, unfortunately operations like this is best implemented in plugins, JsAction was replaced with full support for npm packages via plugins. 

If it is something as generic as this, we'll consider adding it as a default plugin. Can you perhaps expand on your implementation?

**Reply by kemeris2000** [reply]
Hi [Gervwyk](https://github.com/Gervwyk),

this case is quite simple. My app initiate execution of custom script within [netbox DCIM](https://github.com/netbox-community/netbox) by calling API endpoint via AxiosHttp request, response provides another API endpoint (`_request: request.result.url`) for job status.

Response of initiation:
<pre>
    {
        "url": "https://example.domain.com/api/extras/scripts/pdns.pdnscontrol/",
        "id": "pdns.pdnscontrol",
        "module": "pdns",
        "name": "pdnscontrol",
        "description": "Manage DNS zones on PowerDNS Auth servers",
        "vars": {},
        <b>"result": {
            "url": "https://example.domain.com/api/core/jobs/2137/",</b>
            "created": "2023-08-20T18:09:24.013714+03:00",
            "completed": "2023-08-20T18:09:25.522951+03:00",
            "status": {
                "value": "completed",
                "label": "Completed"
            }
        }
    }
</pre>

I need to poll second API endpoint (`https://example.domain.com/api/core/jobs/2137/`) every nth second until predefined max polling attempts reached or polled request response (`_request: request2.status.value`) in array ['completed', 'error']

Response of job status:
<pre>
{
    "id": 2137,
    <b>"url": "https://example.domain.com/api/core/jobs/2137/",</b>
    "display": "21fcd243-dec5-409c-996f-bedfea6a279f",
    "object_type": "extras.scriptmodule",
    "object_id": 51,
    "name": "pdnscontrol",
    <b>"status": {
        "value": "completed",</b>
        "label": "Completed"
    },
    "created": "2023-08-20T18:09:24.013714+03:00",
    "scheduled": null,
    "interval": null,
    "started": "2023-08-20T18:09:24.044465+03:00",
    "completed": "2023-08-20T18:09:25.522951+03:00",
    "data": {
        "log": [],
        "output": "- server: ns3.domain.com (5.5.5.5)\n  stage: zone_mod\n  status: success\n  value: zone aassddff223.com updated successfully\n"
    },
    "job_id": "21fcd243-dec5-409c-996f-bedfea6a279f"
}
</pre>

**Comment by kemeris2000** [comment]
Any news on this plugin? Its very needed missing feature for project I am working on.

**Reply by SamTolmay** [reply]
Hi @kemeris2000 

This something like this should work in V4, it is based on the solution shared for V3:
```js
function PollRequest({ methods: { request }, params: { requestId, interval = 1000 * 60 * 5 } }) {
  setInterval(() => request(requestId), interval);
}

export default PollRequest;
```

You would use it like this:
```yaml
- id: start_poll_request
  type: PollRequest
  params:
    requestId: my_request_id
    delay: 600000
```

**Comment by kemeris2000** [comment]
Thank you for your reply [SamTolmay](https://github.com/SamTolmay), code below works. Is it possible to pass "requestParams" to request in plugin and make PollRequest action synchronous?


```
function PollRequest({ methods: { request }, params: { requestId, delay = 3000, duration = 10000 } }) {
  var intervalId = setInterval(function(){
    request(requestId);
  }, delay);
  setTimeout(function(){
    clearInterval(intervalId);
  }, duration);
}

export default PollRequest;

```
      - id: 6a8df6ce-b987-40de-b854-2a758122b23c
        type: PollRequest
        params:
          requestId: netbox_get_coreJob
          requestParams:
            messages:
              error: false            
          duration: 20000
          delay: 2000
```
